### PR TITLE
Demo: Implement product variants editing in product form

### PR DESCRIPTION
**Pull request by nsams**
_Part of handmade product form that should serve as future best practice and base for Admin Generator._

Add product variants editing using a grid and nested form in the "Variants" tab of the product form.  Demonstrates how to achieve a central save button in combination with a tab that contains a grid.

Also adds a new `StackToolbar` component,  which is a thin wrapper around `Toolbar` to only show the last toolbar inside a stack. Otherwise two toolbars would be shown in the product variants form.

---

<details>
  <summary>Screenshots</summary>
  
Bildschirmfoto vom 2024-02-20 16-29-46

Bildschirmfoto vom 2024-02-20 16-29-55

</details>

### CODE REVIEW

File: .changeset/brave-kiwis-pay.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@comet/admin": minor
+---
+
+Add StackToolbar, a variant of Toolbar component that hides itself in a nested stack
```
Comment (johnnyomair) [inline .changeset/brave-kiwis-pay.md:5]:
```suggestion
Add `StackToolbar`, a variant of `Toolbar` component that hides itself in a nested stack
```

File: packages/admin/admin/src/common/toolbar/StackToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    if (!shouldShowTabBar) {
+        return null;
+    }
+
+    return (
+        <>
+            <Toolbar {...props} />
+        </>
```
Comment (johnnyomair) [inline packages/admin/admin/src/common/toolbar/StackToolbar.tsx:26]:
```suggestion
        <Toolbar {...props} />
```

File: demo/admin/src/products/ProductsGrid.tsx [hunk 3]
```diff
@@ -236,9 +233,12 @@ const productsFragment = gql`
             id
             title
         }
-        variants {
-            image
+        colors {
             name
+            colorHexCode
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:238]: This changed in the base branch, you probably need a rebase

File: demo/admin/src/products/ProductsGrid.tsx [hunk 4]
```diff
@@ -236,9 +233,12 @@ const productsFragment = gql`
...
+        }
+        variants {
+            id
         }
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:242]: This is used in copyData, but removed in onPaste, so a product's variants won't be copied, right? How should this be? Copy the relations or add no variants to the copy?

File: demo/admin/src/products/ProductVariantForm.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    return (
+        <FinalForm<FormValues>
+            apiRef={formApiRef}
+            onSubmit={handleSubmit}
+            mode="edit"
+            initialValues={initialValues}
+            initialValuesEqual={isEqual} //required to compare block data correctly
+            onAfterSubmit={(values, form) => {
+                //don't go back automatically TODO remove this automatismn
+            }}
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductVariantForm.tsx:117]: This has been removed in v6, please remove here

File: demo/admin/src/products/ProductVariantForm.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                    {saveConflict.dialogs}
+                    <MainContent>
+                        <Field
+                            fullWidth
+                            name="name"
+                            component={FinalFormInput}
+                            label={<FormattedMessage id="productVariant.name" defaultMessage="Name" />}
+                        />
+                        <Field name="image" isEqual={isEqual}>
+                            {createFinalFormBlock(rootBlocks.image)}
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductVariantForm.tsx:131]: Out of scope: We should memo this. For instance, when a Rich Text Block is created, every render loses the current cursor position

File: demo/admin/src/products/ProductVariantForm.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            }
+        }
+    };
+
+    if (error) {
+        return <FormattedMessage id="common.error" defaultMessage="An error has occured. Please try again at later" />;
+    }
+
+    if (loading) {
+        return <CircularProgress />;
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductVariantForm.tsx:105]: Please use the Loading component provided by `@comet/admin`

File: demo/admin/src/products/ProductVariantForm.tsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    setTimeout(() => {
+                        stackSwitchApi.activatePage(`edit`, id);
+                    });
+                }
+            }
+        }
+    };
+
+    if (error) {
+        return <FormattedMessage id="common.error" defaultMessage="An error has occured. Please try again at later" />;
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductVariantForm.tsx:101]: Please use the `messages` from `@comet/admin` here

File: demo/admin/src/products/ProductVariantForm.tsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                            {createFinalFormBlock(rootBlocks.image)}
+                        </Field>
+                    </MainContent>
+                </EditPageLayout>
+            )}
+        </FinalForm>
+    );
+}
+
+export default ProductVariantForm;
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductVariantForm.tsx:140]: Please don't use default exports

File: demo/admin/src/products/ProductsGrid.tsx (hunk 3 above)
Comment (nsams) [reply demo/admin/src/products/ProductsGrid.tsx]: hm, this was added in this PR, and I remember changing that. But somehow it got lost....

File: demo/admin/src/products/ProductsGrid.tsx (hunk 4 above)
Comment (nsams) [reply demo/admin/src/products/ProductsGrid.tsx:241]: that depends on the use case, as always :)

let's not copy it for now

**Review by johnnyomair — APPROVED** [review]

**Comment by johnnyomair** [comment]
@jamesricky please review as well, since this will be considered as best practice and will be generated by the Admin Generator.

File: packages/admin/admin/src/common/toolbar/StackToolbar.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    let shouldShowTabBar = true;
+    if (stackApi && stackSwitchApi) {
+        // When inside a Stack show only the last TabBar
+        const ownSwitchIndex = stackSwitchApi.id ? stackApi.switches.findIndex((i) => i.id === stackSwitchApi.id) : -1;
+        const nextSwitchShowsInitialPage = stackApi.switches[ownSwitchIndex + 1] && stackApi.switches[ownSwitchIndex + 1].isInitialPageActive;
+
+        shouldShowTabBar = ownSwitchIndex === stackApi.switches.length - (nextSwitchShowsInitialPage ? 2 : 1);
+    }
+
+    if (!shouldShowTabBar) {
```
Comment (jamesricky) [inline packages/admin/admin/src/common/toolbar/StackToolbar.tsx:19]: The variables/comments with `TabBar` should probably be `Toolbar`, right?

**Review by jamesricky — COMMENTED** [review]
In case it's relevant, the `Toolbar` will be changed quite drastically next sprint, see: COM-13

**Review by thomasdax98 — APPROVED** [review]

**Pull request merged**